pkg/controller/endpointslice/endpointslice_controller.go

```diff
@@ -422,7 +444,11 @@ func (c *Controller) onEndpointSliceUpdate(prevObj, obj interface{}) {
 func (c *Controller) onEndpointSliceDelete(obj interface{}) {
 	endpointSlice := getEndpointSliceFromDeleteAction(obj)
 	if endpointSlice != nil && managedByController(endpointSlice) && c.endpointSliceTracker.Has(endpointSlice) {
-		c.queueServiceForEndpointSlice(endpointSlice)
+		// This returns false if we didn't expect the EndpointSlice to be
+		// deleted. If that is the case, we queue the Service for another sync.
```

Comment:
Hmm.. I'm wondering what happens in case of bug (e.g. we're deleting EndpointSlice but marking tracker incorrectly).
Wouldn't we go into infinite loop here?
Shouldn't we delete the entry from tracker anyway *and* requeue?

[I guess I didn't have time to think deep enough to catch something subtle, but in such case it means it probably requires a more extensive comment.]

Reply:
that is why I'm afraid of not checking the number of requeues here https://github.com/kubernetes/kubernetes/pull/99345/files#r581095710 , I was thinking in possible loops but I couldn't find one, but :shrug:

Reply:
> Shouldn't we delete the entry from tracker anyway and requeue?

Yes good call, I've updated the logic in `Delete` to do that.

> that is why I'm afraid of not checking the number of requeues here

Good point, removed special treatment for this error so num requeues will now be checked.

>Hmm.. I'm wondering what happens in case of bug (e.g. we're deleting EndpointSlice but marking tracker incorrectly).
>Wouldn't we go into infinite loop here?

I think the big risk here is if we called `ExpectDelete` on an EndpointSlice that wasn't actually deleted. That would prevent any future syncs from happening for this Service until that EndpointSlice was actually deleted. We'd run into the same problem if we were to set a generation to a higher value than it actually was persisted as. I don't think it's possible for either of those to happen with the current code, but if it did, it would not be good. 

The previous PR included a [3 second max sync delay to help mitigate these kinds of edge cases](https://github.com/kubernetes/kubernetes/pull/93520/files#diff-8765679aa8ca8cc500a6dc038b8707cbc29ddc2d2f94f1da911e2cd5e4df5816R289). When a service hadn't been synced for that long, it would just ignore cache errors and try anyway. If that or some kind of similar fallback would be valuable, I'm happy to add it in.

Reply:
Let's not add it.

With the changes to Delete, on the next operation ShouldSync will always return true. Which is basically what I wanted.